core/trino-main/src/main/java/io/trino/sql/planner/optimizations/PredicatePushDown.java

```diff
@@ -946,39 +949,87 @@ private InnerJoinPushDownResult processInnerJoin(
+            }
+
+            List<Expression> leftCandidates = extractConjuncts(leftEffectivePredicate).stream()
+                    .filter(EqualityInference::isInferenceCandidate)
+                    .toList();
+
+            List<Expression> leftResiduals = extractConjuncts(leftEffectivePredicate).stream()
+                    .filter(conjunct -> !isInferenceCandidate(conjunct))
+                    .toList();
+
+            List<Expression> rightCandidates = extractConjuncts(rightEffectivePredicate).stream()
+                    .filter(EqualityInference::isInferenceCandidate)
+                    .toList();
+
+            List<Expression> rightResiduals = extractConjuncts(rightEffectivePredicate).stream()
+                    .filter(conjunct -> !isInferenceCandidate(conjunct))
+                    .toList();
 
             ImmutableSet<Symbol> leftScope = ImmutableSet.copyOf(leftSymbols);
             ImmutableSet<Symbol> rightScope = ImmutableSet.copyOf(rightSymbols);
 
-            // Generate equality inferences
-            EqualityInference allInference = new EqualityInference(inheritedPredicate, leftEffectivePredicate, rightEffectivePredicate, joinPredicate);
-            EqualityInference allInferenceWithoutLeftInferred = new EqualityInference(inheritedPredicate, rightEffectivePredicate, joinPredicate);
-            EqualityInference allInferenceWithoutRightInferred = new EqualityInference(inheritedPredicate, leftEffectivePredicate, joinPredicate);
+            EqualityInference allInference = new EqualityInference(
+                    ImmutableList.<Expression>builder()
+                            .addAll(candidates)
+                            .addAll(leftCandidates)
+                            .addAll(rightCandidates)
+                            .build());
+            EqualityInference inferenceWithoutLeft = new EqualityInference(
+                    ImmutableList.<Expression>builder()
+                            .addAll(candidates)
+                            .addAll(rightCandidates)
+                            .build());
+            EqualityInference inferenceWithoutRight = new EqualityInference(
+                    ImmutableList.<Expression>builder()
+                            .addAll(candidates)
+                            .addAll(leftCandidates)
+                            .build());
 
-            // Add equalities from the inference back in
-            leftPushDownConjuncts.addAll(allInferenceWithoutLeftInferred.generateEqualitiesPartitionedBy(leftScope).getScopeEqualities());
-            rightPushDownConjuncts.addAll(allInferenceWithoutRightInferred.generateEqualitiesPartitionedBy(rightScope).getScopeEqualities());
-            joinConjuncts.addAll(allInference.generateEqualitiesPartitionedBy(leftScope).getScopeStraddlingEqualities()); // scope straddling equalities get dropped in as part of the join predicate
+            // See if we can push the right effective predicate to the left side
```

Comment:
this comment seems off.  I expected something like "determine which right conjuncts can be push down the left side"

Reply:
They were carryovers from the previous code